pkg/store/cache/cache.go

```diff
@@ -0,0 +1,325 @@
+	}, []string{"item_type"})
+	c.requests.WithLabelValues(cacheTypePostings)
+	c.requests.WithLabelValues(cacheTypeSeries)
+
+	c.overflow = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_items_overflowed_total",
+		Help: "Total number of items that could not be added to the cache due to being too big.",
+	}, []string{"item_type"})
+	c.overflow.WithLabelValues(cacheTypePostings)
+	c.overflow.WithLabelValues(cacheTypeSeries)
+
+	c.hits = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_hits_total",
+		Help: "Total number of requests to the cache that were a hit.",
+	}, []string{"item_type"})
+	c.hits.WithLabelValues(cacheTypePostings)
+	c.hits.WithLabelValues(cacheTypeSeries)
+
+	c.current = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "thanos_store_index_cache_items",
+		Help: "Current number of items in the index cache.",
+	}, []string{"item_type"})
+	c.current.WithLabelValues(cacheTypePostings)
+	c.current.WithLabelValues(cacheTypeSeries)
+
+	c.currentSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "thanos_store_index_cache_items_size_bytes",
+		Help: "Current byte size of items in the index cache.",
+	}, []string{"item_type"})
+	c.currentSize.WithLabelValues(cacheTypePostings)
+	c.currentSize.WithLabelValues(cacheTypeSeries)
+
+	c.totalCurrentSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "thanos_store_index_cache_total_size_bytes",
+		Help: "Current byte size of items (both value and key) in the index cache.",
+	}, []string{"item_type"})
+	c.totalCurrentSize.WithLabelValues(cacheTypePostings)
+	c.totalCurrentSize.WithLabelValues(cacheTypeSeries)
+
+	if reg != nil {
+		reg.MustRegister(prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+			Name: "thanos_store_index_cache_max_size_bytes",
+			Help: "Maximum number of bytes to be held in the index cache.",
+		}, func() float64 {
+			return float64(c.maxSizeBytes)
+		}))
+		reg.MustRegister(prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+			Name: "thanos_store_index_cache_max_item_size_bytes",
+			Help: "Maximum number of bytes for single entry to be held in the index cache.",
+		}, func() float64 {
+			return float64(c.maxItemSizeBytes)
+		}))
+		reg.MustRegister(c.requests, c.hits, c.added, c.evicted, c.current, c.currentSize, c.totalCurrentSize, c.overflow)
+	}
+
+	// Initialize LRU cache with a high size limit since we will manage evictions ourselves
+	// based on stored size using `RemoveOldest` method.
+	l, err := lru.NewLRU(math.MaxInt64, c.onEvict)
+	if err != nil {
+		return nil, err
+	}
+	c.lru = l
+
+	level.Info(logger).Log(
+		"msg", "created index cache",
+		"maxItemSizeBytes", c.maxItemSizeBytes,
+		"maxSizeBytes", c.maxSizeBytes,
+		"setTimeout", c.setTimeout.String(),
+		"maxItems", "math.MaxInt64",
+	)
+	return c, nil
+}
+
+func (c *IndexCache) onEvict(key, val interface{}) {
+	k := key.(cacheItem).keyType()
+	entrySize := sliceHeaderSize + uint64(len(val.([]byte)))
```

Comment:
Duplicating the entrySize calculation everywhere seems dubious. It's already inconsistent, sometimes using a literal `16` and sometimes `sliceHeaderSize`.

Reply:
sorry, might miss some places ):

Reply:
what do you mean duplicating though?

Reply:
There are three locations where you calculate `entrySize`, all of which must do the same calculation for the cache to work correctly. It might be worth factoring this out, although I guess it might be good enough as is.